File: Subsets.py
#https://leetcode.com/problems/subsets/description/

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        
        def subset(nums,i,result,stack):
            if i==len(nums):
                result.append(list(stack))
            else:
                stack.append(nums[i])
                result=subset(nums,i+1,result,stack)
                stack.pop()
                result=subset(nums,i+1,result,stack)
            return result
        return subset(nums,0,[],[])

        

File: Subsets.cpp
//https://leetcode.com/problems/subsets/description/

class Solution {
public:
    vector<vector<int>> result;
    void helper(vector<int>& nums, vector<int> store, int idx){
        if(idx==nums.size()){
            result.emplace_back(store);
            return;
        }
        //not take
        helper(nums,store,idx+1);
        // take
        store.emplace_back(nums[idx]);
        helper(nums,store,idx+1);
        store.pop_back();
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> store;
        helper(nums,store,0);
        return result;
    }
};
